File: 0025_reverse_k_group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def __init__(self):
        self.count = 0
        self.output = []
        self.stack = []
        self.node = None
        
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        self.recursive(head, k)
        self.output.append(None)
        for i in range(len(self.output)-1):
            self.output[i].next = self.output[i+1]
        return self.output[0]
        
    def recursive(self, head, k):
        if self.count == k:
            self.stack = self.stack[::-1]
            self.output += self.stack
            self.count = 0
            self.stack = []
        if head is None:
            self.output += self.stack
            return
        self.stack.append(head)
        self.count += 1
        self.recursive(head.next, k)
        return

File: 0001-0500/0025_reverseKGroup.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null) {
            return head;
        }

        int step = k;
        ListNode prev = null;
        ListNode curr = head;

        while (curr != null && step > 0) {
            prev = curr;
            curr = curr.next;
            step--;
        }

        prev.next = null;

        if (step > 0) {
            return head;
        } else {
            reverseList(head);
            head.next = reverseKGroup(curr, k);
            return prev;
        }
    }

    public void reverseList(ListNode node) {
        ListNode prev = null;
        ListNode curr = node;
        ListNode next;

        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
    }
}
